app/models/billboard.rb

```diff
@@ -101,6 +103,27 @@ def self.for_display(area:, user_signed_in:, user_id: nil, article: nil, user_ta
     end
   end
 
+  def self.weighted_random_selection(relation)
+    base_query = relation.to_sql
+    random_val = rand.to_f
+
+    query = <<-SQL
+      WITH base AS (#{base_query}),
+      weighted AS (
+        SELECT *, priority_weight,
+        SUM(priority_weight) OVER () AS total_weight,
+        SUM(priority_weight) OVER (ORDER BY id ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS running_weight
+        FROM base
+      )
+      SELECT *, running_weight, ? * total_weight AS random_value FROM weighted
+      WHERE running_weight >= ? * total_weight
+      ORDER BY running_weight ASC
+      LIMIT 1
+    SQL
```

Comment:
I am not _fully_ sure about this query, largely because of the running weight calculation - since the window function is operating on the table ordered by id, _and_ the actual projected set is ordered by running weight in ascending order, does this not mean that older billboards with a high enough priority weight will swamp out newer billboards (even fairly-high-priority-weight ones) that come after them for a good while? That _may_ be the intended behaviour, but I thought I should point it out in case it isn't.

I'm not sure if the spec we currently have would pick that up - the weights in it are all quite small compared to the possibility of up to `10_000`, _plus_ it doesn't have any weights heavier than `5` coming after it.

Reply:
@filleduchaos this is an accurate assessment of the functionality — with _enough_ weight, "new" could get squeezed out. The intended functionality is more small nudges, within an order of magnitude or so of 1.0 — 10k is noted as the max, as I don't think there should be a _hard cap_ that is too small — just in case we do want to ramp up the weight for any reason.

But the general intended use is small adjustments, and if you were to max out the weight to 10k, it would effectively squeeze everything out. So that option is available, but would only be used if that were, in fact, the intention.

Reply:
Oh that makes sense! Yeah I think with small weights and enough billboards, the distribution will lean a bit older but be drawn towards newer high-priority billboards a lot easier